Added Selenium URL scraper as new datasource; modified column filter to allow detailed matching information

dale-wahl opened the pull request:
- Added `selenium_scraper` as a new Search class to be used in creating new datasources.
- Created `url_scraper` datasource which allows a user to scrape a list of urls and up to 5 subpages on the host
- Modified the `column-filter` processor to provide detail output showing which matches were found in a given column (this can mimic Tracker Tracker by searching for substrings within HTML and noting which were found/not found in the HTML)
- updated Search base class to allow `after_search_completed` method; this was necessary to ensure Selenium webdriver and Chrome browser are properly closed and also works with `get_items` generators.
- `validate_url` helper function added to `helpers.py`

@stijn-uva I have tested this and not found any issues, but did want you to particularly review the changes to Search.

dale-wahl commented on backend/abstract/selenium_scraper.py:
```diff
@@ -0,0 +1,208 @@
+        except:
+            pass
+
+    def after_search_completed(self):
```
I actually tried that originally and that was the idea for this method. `BasicProcessor.after_process()` only runs if `process()` completes successfully. I need to ensure Chrome and the webdriver _always_ close, even if it is interrupted or if there is an error. Otherwise you're left with memory hungry Chrome hanging out and, worse off, if another datasource uses Selenium, it will start new instances of both.

But looking at it a bit more, I could override both `after_process` and `remove_files`. I think that will catch `process` success as well as any errors if you don't like the `after_search_completed` method being added.

dale-wahl replied:
Actually, it might make sense to move this `after_search_completed` to a `clean_up` method in `BasicProcessor` that always runs after `process` whether it completes successfully or not. I did it in search since I was thinking there might be other things I would like to do after a search was completed.

It could be where we remove temp files for example (which currently happens in `after_process` and `remove_files` which is in abort and runs on its own if there is an exception). Though if we want to be able to resume datasources, we have to change that anyway since we cannot remove those temp files.

stijn-uva replied:
I really like the former idea - a general-purpose way to add code that will *always* be executed after a processor finishes or crashes would likely be useful in more cases. That's worth adding an extra 'hook'. And as you say it would solve the problem of temporary files being left behind. Resuming datasources is probably best added as a feature in one go, so we don't need to accomodate it yet.

dale-wahl replied:
Ok, added a `clean_up` method to BasicWorker that runs after `work` whether it is successful or not. BasicProcessor then uses that and removes any staging areas. Selenium processor then adds to that method for quit.

I could have added clean_up in the BasicProcessor, but it would actually need to run in each except block and just seemed a bit messier. Plus now any worker can use it.

dale-wahl commented on config.py-example:
```diff
@@ -15,7 +15,8 @@ DATASOURCES = {
     "telegram": {},
-    "twitterv2": {}
+    "twitterv2": {},
+    "url_scraper": {},
```
I'll go ahead and remove it from the defaults since you are concerned, but if a student ran the datasource under that impression, they would, instead of comments/posts, receive HTML code of the facebook login page. That ought to become quickly apparent if it was not already clear in the datasource description. I'm not sure of the real downside to this, but maybe they would email you or open an issue which is time consuming to respond to.

I think an actual issue that would create confusing might be in using the various processors. Some might fail or return strange results as they are designed with text in mind. I already thought it would be worth upgrading this to extract the text (which more naturally fix the `body` field) and pushing the HTML code to a `source_code` field. That just wasn't immediately necessary for my intent.

It may also be worth noting that the defaults only matter to users that install 4CAT directly to their machine or server. If I found 4CAT on Github or from a blog post, I would probably assume that all working features would be readily available and not rely on me searching the source code or Github wiki to activate them. Perhaps a semi-solution might be to add all the datasources to the config with an `active` attribute as True and False. That would couple nicely if we move to a more interactive control panel.

stijn-uva replied:
Yep, that's a good QoL suggestion, and aids discoverability of data sources. And then later there could be a page in the interface where you simply flip switches for specific data sources.

RE: Processors - I wonder if we should explicitly have two categories of 4CAT data sources - social media (i.e. following the thread/post/subject format) and other. I think most processors would mostly be useful to only one of the two, or need explitly defined alternate code paths to deal with both. It would also make it easier to set the right expectations in the interface. It puts this data source sort of beyond the original scope of 4CAT though, so I'm not sure how to best handle that. Maybe we can discuss it in a meeting sometime soon with Sal.

RE: Users would notice if the output is not what they expect - true, but they'd only see that *after* Selenium has spent an hour fetching Facebook pages ;)

dale-wahl commented:
I need to update the setup (Docker and manual). 
Some references I use to get Firefox working properly:
https://takac.dev/example-of-selenium-with-python-on-docker-with-latest-firefox/
https://github.com/mozilla/geckodriver/releases/tag/v0.30.0
Can also remove Google Chrome.

dale-wahl commented:
Updated this branch to work with the new config_manager/database settings. Also updated selenium and made some minor bug fixes. Finally created a separate installation that only runs with the selenium settings are input into 4CAT settings (automatically in Docker when the backend container is restarted).

The only "to-do" left is a quality of life update to the frontend to seperate out these specific datasources since they do not conform to the social media paradigm of other datasources.

dale-wahl commented:
Merged master into this branch and fixed all the conflicts. I had updated the column_filter, but those merges were a mess (both you and I made changes in the master that conflicted) so I left it as is. I may revisit or think up a better way for users to search for specific bugs/trackers.

I tested installation and both the URL and Screenshot processors work. There is a install script for Docker that automatically runs if you add `firefox` to the 4CAT settings (you need to also enable the desired datasources). Right now the install runs when restarting the Docker container. I will try to revisit that and see if I can get it to work with the interactive restart.

dale-wahl commented:
Closing; this will be obsolete with the [extensions PR](https://github.com/digitalmethodsinitiative/4cat/pull/451) and all updates, datasources, and processors have been moved to a [new repository](https://github.com/digitalmethodsinitiative/4cat_web_studies_extensions/tree/main).

I will remove the `tracker_tracker` and `app_stores` branches when the above is merged.
